### PR TITLE
feat: Added Llama2 training examples to trainium-inferentia module

### PR DESCRIPTION
**Pull request by sanjeevrg89**
### What does this PR do?
Showcases how to run distributed training on Trainium using Llama2

🛑 Please open an issue first to discuss any significant work and flesh out details/direction - we would hate for your time to be wasted.
Consult the [CONTRIBUTING](https://github.com/awslabs/data-on-eks/blob/main/CONTRIBUTING.md#contributing-via-pull-requests) guide for submitting pull-requests.

1. Added random name to the EKS cluster instead of hard-coded **trainium-inferentia**. Reason to do this -> Its hard to run multiple EKS clusters with same name and clean up is challenging
2. Added llama2 distributed training examples (Scott Perry and I worked on this module)
3. Automated all the llama2 training examples using shell scripts
4. Automated pre-reqs steps and created a shell script as we are asking users to use EC2 instance or Cloud 9 for Llama2 example
5. Modified Availability zones script under main.tf to properly pick trn1 instances
3. Added MD file for Llama2 distributed training on Trainium
4. Modified outputs.tf, variables.tf, eks.tf and main.tf

### Motivation

Llama2 distributed pre-training example using Trainium on EKS

### More

- [ Yes] Yes, I have tested the PR using my local account setup (Provide any test evidence report under Additional Notes)
- [Yes ] Mandatory for new blueprints. Yes, I have added a example to support my blueprint PR
- [ Yes] Mandatory for new blueprints. Yes, I have updated the `website/docs` or `website/blog` section for this feature
- [ Yes] Yes, I ran `pre-commit run -a` with this PR. Link for installing [pre-commit](https://pre-commit.com/) locally

### For Moderators

- [ ] E2E Test successfully complete before merge?

### CODE REVIEW

File: ai-ml/trainium-inferentia/eks.tf [hunk 1]
```diff
@@ -176,9 +176,9 @@ module "eks" {
         echo "Bootstrap complete. Ready to Go!"
       EOT
 
-      min_size     = 0
-      max_size     = 2
-      desired_size = 0
+      min_size     = 4
+      max_size     = 4
+      desired_size = 4
```
Comment (vara-bonthu) [inline ai-ml/trainium-inferentia/eks.tf:181]: We are using the same node group for multiple LLMs and the min_size of node groups are different to each one. 
Could you change this back to 
```
min_size     = 0
desired_size = 0
```
so that we can ask the users to update the min and desired before running `install.sh`.  Add this in install instructions of the website or blog.

File: ai-ml/trainium-inferentia/main.tf [hunk 2]
```diff
@@ -48,4 +56,20 @@ locals {
     Blueprint  = local.name
     GithubRepo = "github.com/awslabs/data-on-eks"
   }
+} */
+
+data "external" "eks_azs" {
+  program = ["bash", "${path.module}/get_eks_azs.sh"]
+}
```
Comment (vara-bonthu) [inline ai-ml/trainium-inferentia/main.tf:63]: Is the inclusion of this shell script essential? While it's clear that `az3` and `az4` in `us-west-2` are required, this setup might not hold for different regions, causing users to adjust the script accordingly.

My suggestion is to introduce a new `azs` variable immediately after the region variable, where you can directly specify the availability zones for the Trn1 instances.

To guide users, we should detail this requirement on our website and in the blog post, emphasizing the need to update the availability zones in tandem with the region within the variables prior to deploying the solution.

Incorporating a link to the `Trn1` node availability regions in our documentation would be beneficial for users.

File: ai-ml/trainium-inferentia/main.tf [hunk 3]
```diff
@@ -38,8 +46,8 @@ data "aws_ecrpublic_authorization_token" "token" {
   provider = aws.ecr
 }
 
-locals {
-  name   = var.name
+/* locals {
```
Comment (vara-bonthu) [inline ai-ml/trainium-inferentia/main.tf:49]: Could we remove the commented lines from the code?

File: ai-ml/trainium-inferentia/outputs.tf [hunk 4]
```diff
@@ -1,4 +1,9 @@
-output "configure_kubectl" {
+/* output "configure_kubectl" {
   description = "Configure kubectl: make sure you're logged in with the correct AWS profile and run the following command to update your kubeconfig"
   value       = "aws eks --region ${var.region} update-kubeconfig --name ${var.name}"
+} */
```
Comment (vara-bonthu) [inline ai-ml/trainium-inferentia/outputs.tf:4]: what is the reason for this change? 
Could we remove the commented lines from the code?

File: ai-ml/trainium-inferentia/variables.tf [hunk 5]
```diff
@@ -1,6 +1,6 @@
 variable "name" {
   description = "Name of the VPC and EKS Cluster"
-  default     = "trainium-inferentia"
+  default     = "tr-inf"
```
Comment (vara-bonthu) [inline ai-ml/trainium-inferentia/variables.tf:3]: When making code alterations, particularly regarding renaming the `name`, it's crucial to verify that these changes do not disrupt existing docs or blogs. Please ensure you check the current blueprint document for any dependencies before proceeding. you can search the name to find the `trainium-inferentia` to find the occurances.

For example. you need to change this doc to reflect the new name https://github.com/awslabs/data-on-eks/blob/main/website/docs/gen-ai/inference/Llama2.md

and update this doc as well https://github.com/awslabs/data-on-eks/blob/main/website/docs/blueprints/ai-ml/trainium.md by replacing the `trainium` cluster name with new name

Should this be `trn1-inf2` instead?

File: website/docs/gen-ai/training/Llama2.md [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+./install.sh
+```
+
+
+### Verify the resources
+
+Verify the Amazon EKS Cluster
+
+```bash
+aws eks --region us-west-2 describe-cluster --name <cluster-name>
```
Comment (vara-bonthu) [inline website/docs/gen-ai/training/Llama2.md:111]: change the cluster-name to the actual name defined in the varaibles

File: website/docs/gen-ai/training/Llama2.md [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+```
+
+```bash
+# Creates k8s config file to authenticate with EKS
+aws eks --region us-west-2 update-kubeconfig --name <cluster-name>
```
Comment (vara-bonthu) [inline website/docs/gen-ai/training/Llama2.md:116]: same as above

**Review by vara-bonthu — COMMENTED** [review]
I've added a few small suggestions, but aside from that, it looks great! 👍🏼

File: ai-ml/trainium-inferentia/eks.tf (hunk 1 above)
Comment (sanjeevrg89) [reply ai-ml/trainium-inferentia/eks.tf:181]: I did change this to variables var.min_size, var.max_size, var.desired_size. Set default to zero. Did not modify the code for core node group

Comment (sanjeevrg89) [reply ai-ml/trainium-inferentia/eks.tf:181]: Who is running this can do something like this: export TF_VAR_min_size=x

Comment (sanjeevrg89) [reply ai-ml/trainium-inferentia/eks.tf:181]: > We are using the same node group for multiple LLMs and the min_size of node groups are different to each one. Could you change this back to
> 
> ```
> min_size     = 0
> desired_size = 0
> ```
> 
> so that we can ask the users to update the min and desired before running `install.sh`. Add this in install instructions of the website or blog.

Yes I changed this back to 0

File: ai-ml/trainium-inferentia/outputs.tf (hunk 4 above)
Comment (sanjeevrg89) [reply ai-ml/trainium-inferentia/outputs.tf:4]: I was having issues using hardcoded eks cluster name. If I had to deploy another set of cluster I had issues because KMS would not allow creation of another AWS managed key with the same name. That is why I have used random string to the cluster name thats hardcoded.

Comment (sanjeevrg89) [reply ai-ml/trainium-inferentia/outputs.tf:4]: I will remove the commented lines

File: ai-ml/trainium-inferentia/variables.tf (hunk 5 above)
Comment (sanjeevrg89) [reply ai-ml/trainium-inferentia/variables.tf:3]: I will update the name to reflect trn1-inf2

File: website/docs/gen-ai/training/Llama2.md (hunk 6 above)
Comment (sanjeevrg89) [reply website/docs/gen-ai/training/Llama2.md:111]: The reason I cannot use the same name is I am appending a random string as discussed above

File: ai-ml/trainium-inferentia/variables.tf (hunk 5 above)
Comment (sanjeevrg89) [reply ai-ml/trainium-inferentia/variables.tf:3]: **NOTE:** Replace [cluster-name] with your actual EKS cluster name

Added this to the documentation because we cannot hardcode cluster name in the docs as we have added random string to the cluster name

File: ai-ml/trainium-inferentia/main.tf (hunk 2 above)
Comment (sanjeevrg89) [reply ai-ml/trainium-inferentia/main.tf]: Exporting the region variable and running that script as part of terraform installation is ideal than asking users to specify Availability zones.

File: ai-ml/trainium-inferentia/main.tf [hunk 8]
```diff
@@ -39,7 +47,7 @@ data "aws_ecrpublic_authorization_token" "token" {
 }
 
 locals {
-  name   = var.name
+  name   = "${var.name}-${random_string.this.result}"
```
Comment (vara-bonthu) [inline ai-ml/trainium-inferentia/main.tf:50]: remove this and put the name back to `var.name` and keep the name as `trainium-inferentia`.

File: ai-ml/trainium-inferentia/main.tf [hunk 9]
```diff
@@ -30,6 +30,14 @@ provider "kubectl" {
...
 }
 
+resource "random_string" "this" {
+  length  = 5
+  special = false
+  upper   = false
+  lower   = true
+  numeric = true
+}
+
```
Comment (vara-bonthu) [inline ai-ml/trainium-inferentia/main.tf:40]: remove this one

File: ai-ml/trainium-inferentia/get_eks_azs.sh [hunk 10]
```diff
@@ -0,0 +1,43 @@
+#!/bin/bash
```
Comment (vara-bonthu) [inline ai-ml/trainium-inferentia/get_eks_azs.sh:1]: remove this file

File: ai-ml/trainium-inferentia/variables.tf [hunk 11]
```diff
@@ -1,10 +1,11 @@
 variable "name" {
   description = "Name of the VPC and EKS Cluster"
-  default     = "trainium-inferentia"
```
Comment (vara-bonthu) [inline ai-ml/trainium-inferentia/variables.tf:3]: keep the same name is its being used in multiple blueprints

File: ai-ml/trainium-inferentia/variables.tf [hunk 12]
```diff
@@ -49,3 +50,76 @@ variable "enable_mpi_operator" {
...
   default     = false
 }
+
+variable "trn1_32xl_min_size" {
+  description = "trn1 Worker node minimum size"
+  type = number
+  default = 0
+}
+
+variable "trn1_32xl_max_size" {
```
Comment (vara-bonthu) [inline ai-ml/trainium-inferentia/variables.tf:60]: remove all the max variables and this can be hardcoded

File: ai-ml/trainium-inferentia/variables.tf [hunk 13]
```diff
@@ -49,3 +50,76 @@ variable "enable_mpi_operator" {
...
+  default = 0
+}
+
+variable "trn1n_32xl_min_size" {
+  description = "Worker node minimum size"
+  type = number
+  default = 0
+}
+
+variable "trn1n_32xl_max_size" {
```
Comment (vara-bonthu) [inline ai-ml/trainium-inferentia/variables.tf:78]: remove all the max variables and this can be hardcoded

File: ai-ml/trainium-inferentia/variables.tf [hunk 14]
```diff
@@ -49,3 +50,76 @@ variable "enable_mpi_operator" {
...
+  default = 0
+}
+
+variable "inf2_24xl_min_size" {
+  description = "Worker node minimum size"
+  type = number
+  default = 0
+}
+
+variable "inf2_24xl_max_size" {
```
Comment (vara-bonthu) [inline ai-ml/trainium-inferentia/variables.tf:96]: remove all the max variables and this can be hardcoded

File: ai-ml/trainium-inferentia/variables.tf [hunk 15]
```diff
@@ -49,3 +50,76 @@ variable "enable_mpi_operator" {
...
+  default = 0
+}
+
+variable "inf2_48xl_min_size" {
+  description = "Worker node minimum size"
+  type = number
+  default = 0
+}
+
+variable "inf2_48xl_max_size" {
```
Comment (vara-bonthu) [inline ai-ml/trainium-inferentia/variables.tf:114]: remove all the max variables and this can be hardcoded

File: website/docs/gen-ai/inference/Llama2.md [hunk 16]
```diff
@@ -114,13 +114,15 @@ cd data-on-eks/ai-ml/trainium-inferentia/ && chmod +x install.sh
 
 Verify the Amazon EKS Cluster
 
+**NOTE:** Replace [cluster-name] with your actual EKS cluster name
+
 ```bash
-aws eks --region us-west-2 describe-cluster --name trainium-inferentia
```
Comment (vara-bonthu) [inline website/docs/gen-ai/inference/Llama2.md:118]: put this change back

File: website/docs/gen-ai/inference/Llama2.md [hunk 17]
```diff
@@ -114,13 +114,15 @@ cd data-on-eks/ai-ml/trainium-inferentia/ && chmod +x install.sh
...
+aws eks --region us-west-2 describe-cluster --name [cluster-name]
 ```
 
 ```bash
 # Creates k8s config file to authenticate with EKS
-aws eks --region us-west-2 update-kubeconfig --name trainium-inferentia
```
Comment (vara-bonthu) [inline website/docs/gen-ai/inference/Llama2.md:123]: put this change back

File: website/docs/gen-ai/inference/Llama2.md [hunk 18]
```diff
@@ -148,7 +150,7 @@ Users can also modify the Dockerfile to suit their specific requirements and pus
 
 **Ensure the cluster is configured locally**
 ```bash
-aws eks --region us-west-2 update-kubeconfig --name trainium-inferentia
```
Comment (vara-bonthu) [inline website/docs/gen-ai/inference/Llama2.md:151]: put this change back

Comment (sanjeevrg89) [reply website/docs/gen-ai/inference/Llama2.md:151]: Reverted

File: website/docs/gen-ai/inference/Llama2.md (hunk 17 above)
Comment (sanjeevrg89) [reply website/docs/gen-ai/inference/Llama2.md:123]: reverted

File: ai-ml/trainium-inferentia/variables.tf (hunk 11 above)
Comment (sanjeevrg89) [reply ai-ml/trainium-inferentia/variables.tf:3]: reverted

File: ai-ml/trainium-inferentia/get_eks_azs.sh (hunk 10 above)
Comment (sanjeevrg89) [reply ai-ml/trainium-inferentia/get_eks_azs.sh:1]: removed this file

File: ai-ml/trainium-inferentia/main.tf (hunk 9 above)
Comment (sanjeevrg89) [reply ai-ml/trainium-inferentia/main.tf:40]: removed the random string

File: ai-ml/trainium-inferentia/main.tf (hunk 8 above)
Comment (sanjeevrg89) [reply ai-ml/trainium-inferentia/main.tf:50]: done

**Review by vara-bonthu — APPROVED** [review]
LGTM 👍🏼
